### ISSUE TITLE
[Bug]: libx11 1.8.2 XLookupString gives 0 bytes for non ascii chars

### ISSUE THREAD

**Issue by jidhub** [termux/termux-packages#13362]
### Problem description

libx11, which is used by xterm (from aterm) and xorg-xev, returns empty strings for non-ASCII keypresses.

This is a problem for me to enter french language, using lots of à é è and some ç ô ê â ...

### What steps will reproduce the bug?

I opened a terminal on termux with:

Xvfb -screen 0 2208x1768x24 &
DISPLAY=:0 x11vnc -rfbauth ~/p/r/.vnc/passwd -shared -forever
DISPLAY=:0 xterm

Then I use bVNC [1].

Then I focus the xterm, execute vim, then type i to enter insertmode. Then all ascii chars are printed as normal, but non-ascii like € are ignored by libx11. I can make € but only with the following combination (vim digraphs): control-O e =

If, instead of xterm, I use xev[2], when I press euro on the android keyboard in bVNC, I am shown: 

KeyPress event, serial 24, synthetic NO, window 0xc00001,
    root 0x43, subw 0x0, time 1376978491, (123,123), root:(125,125),
    state 0x0, keycode 103 (keysym 0x20ac, EuroSign), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False

"0 bytes" above is why vim is even not notified of anything happening.

[1] https://play.google.com/store/apps/details?id=com.iiordanov.freebVNC&hl=fr
[2] installed by: pkg install xorg-xev

### What is the expected behavior?

"3 bytes: €" should be shown instead of "0 bytes:".

I also show what pressing euro in the android keyboard shows in the window (X11Forwarding) inside bVNC, after the following command: DISPLAY=:0 ssh -Y jidhub@linux_hostname xev

KeyPress event, serial 28, synthetic NO, window 0xc00001,
    root 0x43, subw 0x0, time 1376947528, (123,123), root:(125,125),
    state 0x0, keycode 103 (keysym 0x20ac, EuroSign), same_screen YES,
    XLookupString gives 3 bytes: (e2 82 ac) "€"
    XmbLookupString gives 3 bytes: (e2 82 ac) "€"
    XFilterEvent returns: False

This is because the xev from my linux box, coming to termux's Xvfb by using X11forwarding, is using a libX11 which is configured to show correctly the euro sign.

### System information

termux-info:
Termux Variables:                                      
TERMUX_APP_PACKAGE_MANAGER=apt                      
TERMUX_MAIN_PACKAGE_FORMAT=debian             
TERMUX_VERSION=0.118.0                                 
Packages CPU architecture:                                                                                     
aarch64                                                                                                        
Subscribed repositories:                               
# sources.list       
deb https://mirrors.sahilister.in/termux/termux-main stable main                                               
# x11-repo (sources.list.d/x11.list)                   
deb https://mirrors.sahilister.in/termux/termux-x11 x11 main                                                   
# root-repo (sources.list.d/root.list)                 
deb https://mirrors.sahilister.in/termux/termux-root root stable                                               Updatable packages:
android-tools/stable 31.0.3p1-7 aarch64 [upgradable from: 31.0.3p1-6]                                          
asciidoctor/stable,stable 2.0.18 all [upgradable from: 2.0.17]                                                 
binutils-is-llvm/stable 0.3-1 aarch64 [upgradable from: 0.3]                                                   
busybox/stable 1.35.0-3 aarch64 [upgradable from: 1.35.0-2]
clang/stable 15.0.5 aarch64 [upgradable from: 15.0.4]
cmake/stable 3.24.3 aarch64 [upgradable from: 3.24.2-1]                                                        
gettext/stable 0.21.1 aarch64 [upgradable from: 0.21-7]                                                        
inetutils/stable 2.4 aarch64 [upgradable from: 2.3-2]  
krb5/stable 1.20.1 aarch64 [upgradable from: 1.20-1]   
libcompiler-rt/stable 15.0.5 aarch64 [upgradable from: 15.0.4]                                                 
libedit/stable 20221030-3.1-0 aarch64 [upgradable from: 20210910-3.1-0]                                        
libheif/stable 1.14.0 aarch64 [upgradable from: 1.13.0]                                                        
libllvm/stable 15.0.5 aarch64 [upgradable from: 15.0.4]                                 
liblzma/stable 5.2.8 aarch64 [upgradable from: 5.2.7]                                                          
libmpfr/stable 4.1.1-1 aarch64 [upgradable from: 4.1.0-1]                   
libnghttp2/stable 1.51.0 aarch64 [upgradable from: 1.50.0]                    
libx11/stable 1.8.2 aarch64 [upgradable from: 1.8.2]                                                           
libxdmcp/stable 1.1.4 aarch64 [upgradable from: 1.1.3-10]   
libxpm/x11 3.5.14 aarch64 [upgradable from: 3.5.13-18] 
lld/stable 15.0.5 aarch64 [upgradable from: 15.0.4]    
llvm/stable 15.0.5 aarch64 [upgradable from: 15.0.4]
nano/stable 7.0 aarch64 [upgradable from: 6.3]
pcre2/stable 10.40-1 aarch64 [upgradable from: 10.40]  
termux-tools/stable,stable 1.31.1 all [upgradable from: 1.31.0]                                                
tigervnc/x11 1.10.1-28 aarch64 [upgradable from: 1.10.1-27]                                                    
xz-utils/stable 5.2.8 aarch64 [upgradable from: 5.2.7] 
termux-tools version:
1.31.0                                                                                                         
Android version:                                       
12                                                                                                             
Kernel build information:                              
Linux localhost 5.4.147-qgki-24879461-abF926BXXS2CVIG #1 SMP PREEMPT Mon Sep 26 12:42:02 KST 2022 aarch64 Android                 
Device manufacturer:                                                                                           
samsung                                                                                                        
Device model:                                                                                                  
SM-F926B

**Comment by jidhub** [comment]
How can I add the label https://github.com/termux/termux-packages/labels/x11 ?

**Comment by jidhub** [comment]
I attach gdb logs.

[gdb.txt](https://github.com/termux/termux-packages/files/10069745/gdb.txt)

Here is how I added symbols and source code for gdb.txt, the command below are *not* *needed* to reproduce the above bug:

In termux:
     sshd -p 64000
     run https://github.com/termux/termux-packages/blob/master/scripts/setup-termux.sh 
     ./build-package.sh -I -d xorg-xev

On linux_hostname, run `https://github.com/termux/termux-packages/blob/master/scripts/run-docker.sh`
This gives a shell in which I type:
    cd ~ ; cp -a termux-packages u ; cd u ; ./build-package.sh -d xorg-xev
    scp -rpP64000 ~/.termux-build output android12:

In termux:
     apt install ~/output/libx11-dbg_1.8.2_aarch64.deb
     grep --devices=skip -R -lZ '/home/builder/.termux-build' /data/data/com.termux/files/usr | xargs -r0n 1 sh -c 'LANG= sed -i -e s:/home/builder/.termux-build:/data/data/com.termux/hmbld:g "$1" ; true' sh
     ln -s /data/data/com.termux/files/home/.termux-build  /data/data/com.termux/hmbld
     DISPLAY=:0 gdb xev
     set pagination off
     break XLookupString
     run
     python [gdb.execute('s') for x in range(19600)]

**Comment by jidhub** [comment]
Thanks カカシ @licy183 for adding the label x11!
Github shows the button "Show your support for https://github.com/termux by sponsoring them" below: is that the right way to promote my bug report, or is there some better "bounty" mechanism for termux developpers ?

**Comment by xtkoba** [comment]
FWIW, this reminds me of #10122. `XSupportsLocale()` returns `False` on Termux.

```
$ cat xsupportslocale.c
int XSupportsLocale(void);
int main()
{
  return !XSupportsLocale();
}
$ clang xsupportslocale.c -lX11
$ ./a.out; echo $?
1
```

So the first step towards resolving this issue should be to somehow make `XSupportsLocale()` return `True`.

**Comment by jidhub** [comment]
> So the first step towards resolving this issue should be to somehow make `XSupportsLocale()` return `True`.

Will do, see https://github.com/termux/termux-packages/compare/master...jidhub:termux-packages:master 

Now I got a 503:
scripts/update-docker.sh:7 docker pull ghcr.io/termux/package-builder
Using default tag: latest
Error response from daemon: received unexpected HTTP status: 503 Service Unavailable

Unable to find image 'ghcr.io/termux/package-builder:latest' locally
docker: Error response from daemon: received unexpected HTTP status: 503 Service Unavailable.
See 'docker run --help'.

Is there another server thant ghcr.io that I can use ?

**Comment by jidhub** [comment]
@xtkoba should have I notified you in above message ?

**Comment by xtkoba** [comment]
> > So the first step towards resolving this issue should be to somehow make `XSupportsLocale()` return `True`.
> 
> Will do, see [jidhub@26d3a78](https://github.com/jidhub/termux-packages/commit/26d3a785497a7354eb21edd653b39ca6792a4a04)

Is this a joke? Or are you serious?

**Comment by jidhub** [comment]
@xtkoba I don't quite understand about the funny part, but I am right now installing the modified .deb (and, right now, figuring out how to tell apt install to reinstall a package).

What was more appropriate to do ?

**Comment by xtkoba** [comment]
I did never mean to define `XSupportsLocale()` as

```c
Bool XSupportsLocale(void) { return True; }
```

Rather I meant Xlib should think it is capable of operating under the current locale (`C.UTF-8`).

**Comment by jidhub** [comment]
@xtkoba > Rather I meant Xlib should think it is capable of operating under the current locale (`C.UTF-8`).

Ok understood. This I cannot guess how to do that, except if you show me a similar modification making some open source think it is capable of operating under the current locale (C.UTF-8).

If you can provide the patch, I'll happily test it though.

**Comment by jidhub** [comment]
@xtkoba Ok you are right, this doesn't fix my problem:

"KeyPress event, serial 24, synthetic NO, window 0xc00001,
    root 0x43, subw 0x0, time 1529829333, (103,170), root:(105,172),
    state 0x0, keycode 103 (keysym 0x20ac, EuroSign), same_screen YES,
    XLookupString gives 0 bytes: 
    XFilterEvent returns: False"

**Comment by xtkoba** [comment]
This is very specific to Xlib. A typical program uses `setlocale()` and/or `nl_langinfo()` to determine the current locale and usually they work fine on Termux (we use a custom `nl_langinfo()` implementation to coerce the locale to `C.UTF-8`).

**Comment by jidhub** [comment]
I see I cannot compile aterm in the docker, so I created the PR https://github.com/termux/termux-packages/pull/13501

**Comment by jidhub** [comment]
@xtkoba I just also tested with aterm:
    gdb aterm
    break XSupportsLocale
    run

But XSupportsLocale is never called.

**Comment by xtkoba** [comment]
Another criteria would be whether `XSetLocaleModifiers("")` returns `NULL`:

```
$ cat xsetlocalemodifiers.c
char *XSetLocaleModifiers(char *);
int main()
{
  return XSetLocaleModifiers("") == 0;
}
$ clang xsetlocalemodifiers.c -lX11
$ ./a.out; echo $?
1
```

**Comment by jidhub** [comment]
> Another criteri~a~on would be whether `XSetLocaleModifiers("")` returns `NULL`:
> 
> ```
> $ cat xsetlocalemodifiers.c
> char *XSetLocaleModifiers(char *);
> int main()
> {
>   return XSetLocaleModifiers("") == 0;
> }
> $ clang xsetlocalemodifiers.c -lX11
> $ ./a.out; echo $?
> 1
> ```

It returns NULL. And this is because it calls _XlcCurrentLC which call _XOpenLC that both return NULL.

**Comment by jidhub** [comment]
This can be seen in the gdb trace of aterm attached here:
[20221127_android.txt](https://github.com/termux/termux-packages/files/10146864/20221127_android.txt)

**Comment by jidhub** [comment]
If you want the source files, untar https://www.x.org/releases/individual/lib/libX11-1.8.2.tar.gz

**Comment by Iey4iej3** [comment]
I am not really sure how this issue is related to locales. I mean, for example, you press é on a French keyboard with French layout, even if `LC_ALL=en_US.UTF-8`, or `LANG=C`, you still get `é`. I think that this is more about keyboard layouts rather than locales.

**Comment by xtkoba** [comment]
Should be fixed with `libx11` version `1.8.4-1`. Thanks.
